keps/sig-architecture/20190412-conformance-behaviors.md

```diff
@@ -0,0 +1,342 @@
+
+        // Suites is a list containing each suite of behaviors for this area.
+        Suites []Suite `json:"suites,omitempty"`
+}
+
+type Suite struct {
+        // Suite is the name of this suite.
+        Suite       string     `json:"suite,omitempty"`
+
+        // Level is `Conformance` or `Validation`.
+        Level       string     `json:"level,omitempty"`
+
+        // Description is a human-friendly description of this suite, possibly
+        // for inclusion in the conformance reports.
+        Description string     `json:"description,omitempty"`
+
+        // Behaviors is the list of specific behaviors that are part of this
+        // suite.
+        Behaviors   []Behavior `json:"behaviors,omitempty"`
+}
+
+type Behavior struct {
+        // Id is a unique identifier for this behavior, and will be used to tie
+        // tests and their results back to this behavior. For example, a
+        // behavior describing the defaulting of the PodSpec nodeSelector might
+        // have an id like `pods/spec/nodeSelector/default`.
+        Id          string `json:"id,omitempty"`
+
+        // ApiObject is the object whose behavior is being described. In
+        // particular, in generated behaviors, this is the object to which
+        // ApiField belongs. For example, `core.v1.PodSpec` or
+        // `core.v1.EnvFromSource`.
+        ApiObject   string `json:"apiObject,omitempty"`
+
+        // ApiField is filled out for generated tests that are testing the
+        // behavior associated with a particular field. For example, if
+        // ApiObject is `core.v1.PodSpec`, this could be `nodeSelector`.
+        ApiField    string `json:"apiField,omitempty"`
+
+        // ApiType is the data type of the field; for example, `string`.
+        ApiType     string `json:"apiType,omitempty"`
+
+        // Generated is set to `true` if this entry was generated by tooling
+        // rather than hand-written.
+        Generated   bool   `json:"generated,omitempty"`
+
+        // Description specifies the behavior. For those generated from fields,
+        // this will identify if the behavior in question is for defaulting,
+        // setting at creation time, or updating, along with the API schema field
+        // description.
+        Description string `json:"description,omitempty"`
+}
+```
+
+### Behavior and Test Generation Tooling
+
+Some sets of behaviors may be tested in a similar, mechanical way. Basic CRUD
+operations, including updates to specific fields and constraints on immutable
+fields, operate in a similar manner across all API resources. Given this, it is
+feasible to automate the creation of simple tests for these behaviors, along
+with the behavior descriptions in the `spec.yaml`. In some cases a complete test
+may not be easy to generate, but a skeleton may be created that can be be
+converted into a valid test with minimal effort.
+
+For these tests, the input is a set of manifests that are applied to the
```

Comment:
Doesn't this put is back into a similar position as we are today? How will we know how much "coverage" we have when it comes to these, more complex, behavioral tests? Arent those nearly just as manual as the tests today?

I do like the simplicity of having a standard approach (apply manifest, wait for condition, end of story) but what does that mean for all the other complex stuff that we currently do in tests?
 - non-manifest actions, like tearing down a node or checking resources before/after an action
 - larger cycles of apply changes/testX/more changes/testY/etc/etc that can't be reduced

It just seems like this only applies to a small subset of tests of the form (createX testY); is that the intent?

Reply:
The idea is to get tooling to handle the simple cases that we can in an automated-as-much-as-possible way. We'll never be able to build tooling to handle the complex cases. We can have tooling that helps scaffold the complex cases, which should help speed development.

The goal here is not to automate away the entire process. It's just to boost the velocity. In v1.15, we have 221 tests IIRC; in v1.9 is was something like 159. By my estimates we need > 2500 tests to cover the basic functions we need (even without new features). That means we should finish around in about 40 releases at the current pace. Ten years is a bit long to wait :).

So, general approach (this should be clarified in the KEP.
* For generating the list of behaviors:
  * Simple behaviors are generated through tooling.
  * Complex behaviors are defined by the SIGs.
* For test creation:
  * Simple behaviors use a basic set of mechanical test scaffolding with fill-in-the-blank validation as needed. 
  * Complex behaviors are hand-written, possibly (hopefully) using some validation primitives from the simple behaviors.